genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.Files;
+import com.netflix.genie.agent.cli.ArgumentDelegates;
+import com.netflix.genie.agent.execution.exceptions.DownloadException;
+import com.netflix.genie.agent.execution.locks.Lock;
+import com.netflix.genie.agent.execution.locks.impl.LockingFactory;
+import com.netflix.genie.agent.execution.services.FetchingCacheService;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+import org.springframework.util.DigestUtils;
+import org.springframework.util.FileCopyUtils;
+import org.springframework.util.FileSystemUtils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.RandomAccessFile;
+import java.net.URI;
+import java.nio.channels.FileChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * A file system based cache on local disk that uses URIs as keys and transparently downloads
+ * missing entries. It uses OS level locks to accomplish the following
+ * <p>
+ * Allows multiple agents reading/writing/deleting in the same cache folder
+ * without corrupting the cache
+ * Downloads new versions of resources once they are available remotely
+ * Deletes the old versions of resources once a new version is successfully downloaded
+ * Recovers from partial downloads of resources in case an agent gets killed in the middle of a download
+ * or a download fails for any other reason
+ * <p>
+ * Does NOT handle concurrency within the same agent (not an issue at the moment)
+ * <p>
+ * Cache structure on local disk
+ * Each resource has a hash to represent it. The version of the resource is the remote last modified
+ * timestamp. For each resource and version, a lock file is created. Each process takes a lock on
+ * this file and then downloads the resource to a data.tmp file. Once the download is complete
+ * data.tmp is atomically moved to data.
+ * For e.g for a hash value of 6d331abc92bc8244bc5d41e2107f303a and last modified = 1525456404
+ * cache entries would like the following
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/data
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/lock
+ * <p>
+ * Deletion of older versions
```

Comment:
Good documentation to here but it doesn't explain how the deletion is handled.

Reply:
Thanks for pointing. Forgot to add